---
title: "Prepare Data for Calibrating dDNA with Mocks"
output: 
  html_document:
    toc: yes
    toc_float: yes
---

**Date written: May 20 2023**

**Author: M Fisher**

**Last Run: `r Sys.Date()`**


## Description

Prepare the metabarcoding data for use in the quantitative model run using `zoid` R package. This includes adding predator (*Carcinus meanas*) DNA back into the taxonomy data set, cross-referencing species between the mock community and sample (stomach content) data, and formatting the data frames to match the input format required by `zoid`.

<br>

*Run notes*
  - re-run Mar 18 2024 for documentation. No changes to final data from previous runs of zoid.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(magrittr)
library(here)
library(ggplot2)
library(ggrepel)

source(here('../','R','format_metabarcoding_data_forQM_RPK.R'))
```


User inputs:
```{r}
# Mock communities #
## input DNA proportions
prop_in_file <- here('../','data','metadata','egc_mcomm_dna_inputs.csv')
## filtered BLAST data, with taxonomy
tax_file <- here('../','data','results','allRuns_BF3_filtered_FINAL_unique_taxa.csv') # also for sample data
mock_tax_file <- here('../','data','results','allRuns_BF3_filtered_unique_taxa.csv')


# Sample data #
## sequencing runs in this data
miseqruns <- c(2,3,4,5)
## blast directory (one above the specific run directories)
blastdir <- '../data/blast'
## suffix of the unfiltered blast data (should be the same for every run, if using my scripts)
blast_suffix <- '_bf3_sample_CEGblast_lca_taxonomy.csv'
## suffix of the fasta file output from blast
fasta_suffix <- c('hash_key_clean_blast_2023-04-26.fasta')

# Are you running this for the first time, or knitting it? #
knitting <- TRUE
```




## Data

sampling run metadata
```{r}
metadat <- read_csv(here('../','data','metadata',paste0('runs',min(miseqruns),'-',max(miseqruns),'_BF3_run_metadata.csv')))
```

### BLAST data, filtered

should contain information for both the mock communities and the samples.

Read in the filtered blast data with taxonomy
```{r}
sample.tax <- read_csv(tax_file)  ## FINAL filtered file (includes manual filtering)
mc.tax <- read_csv(mock_tax_file) ## does not include manual filtering
```

Clean up the mock community data frame, and make sure that all sample names use a '_' instead of a '-'
```{r}
mc.tax <- mc.tax %>% dplyr::select(-c(site_month, type, mc_rm,egc_id)) %>%
  filter(grepl("MC",sample)) %>% distinct()

unique(mc.tax$sample)

mc.tax %<>% mutate(sample=str_replace(sample,"-","_"))

unique(mc.tax$sample)
```


### EGC data

Read in the unfiltered blast data, to pull the green crab (predator IDs) that were filtered out.
```{r}
for(r in miseqruns){
  if(r==miseqruns[1]){
    tax.unfiltered <- read_csv(here(blastdir,paste0('run',r,blast_suffix))) %>% mutate(sample_id=as.character(sample_id))
  } else{
    tax.unfiltered <- tax.unfiltered %>% 
      bind_rows(read_csv(here(here(blastdir,paste0('run',r,blast_suffix)))) %>% mutate(sample_id=as.character(sample_id)))
  }
}
tax.egc <- tax.unfiltered %>% filter(taxon %in% c("Carcinus maenas","Carcinus"))

tax.egc %<>%  mutate(MiSeqRun=ifelse(dataset=="run_2",2,
                                    ifelse(dataset=="run_3",3,
                                           ifelse(dataset=="run_4",4,
                                                  ifelse(dataset=="run_5",5,NA)))))
```

**Mock community** EGC tax data (will only show mocks with EGC in them, A and D)
```{r}
mc.egc <- tax.egc %>% 
  left_join(mc.tax %>% dplyr::select(Sample_name,MiSeqRun,sample,tech) %>% distinct(),
            by=c("Sample_name","MiSeqRun")) %>%
  filter(!is.na(sample)) %>% dplyr::select(-sample_id)
unique(mc.egc$sample)
```

**Sample** EGC tax data
```{r}
sample.egc <- tax.egc %>% 
  anti_join(mc.egc,by=c("Sample_name","MiSeqRun")) %>% # remove mock community samples
  left_join(metadat %>% dplyr::select(-sample_label, -source), by=c("Sample_name"="sample_name","MiSeqRun"="run"))
                                 
length(unique(sample.egc$sample)) # more than in the final data file because predator DNA was filtered out for other analysis
```

### Mock inputs

Proportions of DNA per species that were put into the mock communities
```{r}
indat <- read_csv(prop_in_file)

## all mocks prepared in the metabarcoding data?
all(unique(indat$sample) %in% unique(mc.tax$sample))

## example:
indat %>% filter(sample=="MC_C") %>% dplyr::select("sample","sample_label","species","dna_ng")
```


### BLAST fasta files

use this to check percent identities in the mocks data, as needed
```{r}
for(i in seq(1,length(miseqruns))){
  r=miseqruns[[i]]
  if(r==miseqruns[1]){
    fastas <- read_delim(here(blastdir, paste0("run",r,"_",fasta_suffix)), col_names=FALSE, col_types=rep("c",times=18)) %>% mutate(MiSeqRun=r)
  } else{
    fastas %<>% bind_rows(read_delim(here(blastdir, paste0("run",r,"_",fasta_suffix)), col_names=FALSE, col_types=rep("c",times=18))
                            %>% mutate(MiSeqRun=r,X15=as.character(X15)))
  }
}
colnames(fastas) <- c("qseqid", "sseqid", "sacc", "pident", "length",
                        "mismatch", "gapopen", "qcovus", "qstart", "qend", "sstart", "send", "evalue", "bitscore", "staxids", "qlen", "sscinames", "sseq","MiSeqRun")
```



## Mock Data prep

### merge all hashes

Add in the green crab hashes to the filtered taxonomy
```{r}
mc.tax %<>% bind_rows(mc.egc %>% dplyr::select(all_of(colnames(mc.tax))))

any(is.na(mc.tax$sample))
```

### manual filter of taxa

Double check the taxonomic IDs in the dDNA data. What are all of the unique IDs in the dDNA data? 
```{r}
unique(mc.tax$taxon)
```

Which species from the tissue samples don't show up in the dDNA data?
```{r}
unique(indat$species)[which(!(unique(indat$species) %in% mc.tax$species))]
```

*Mya arenaria* was only ID'd to *Mya*. 
*Cottus asper* was only ID'd to *Cottus*. There may be some other IDs (Dasycottus setiger) that result from poor match between the primer set and the species' DNA. But we should throw these out.
*Zostera marina* was only ID'd to *Zosteraceae*.
*Crangon crangon* seems to be a mis-ID of *Crangon franciscorum*. In the sample data, we backed all *Crangon crangon* IDs out to *Crangon*. 

We don't have any *Mytilis trossulus* or *Mytilis* sequences.

Check the blast percent identity of the *Crangon crangon* ID: 
```{r}
check_pident <- filter(mc.tax, taxon=="Crangon crangon") %>% dplyr::select(sample,tech,Hash,MiSeqRun,taxon) %>%
  left_join(fastas, by=c("Hash"="qseqid","MiSeqRun")) %>%
  group_by(sample,tech,taxon) %>% summarise(n_hashes=length(unique(Hash)),max_pi=max(pident),min_pi=min(pident),mean_pi=mean(pident))

check_pident
```

By the same rules as applied to the sample data, we should back this ID out to *Crangon*. 

```{r eval=FALSE}
```


**Get rid of DNA that was not intentionally put into the mock communities**
```{r}
mc.tax.filter <- mc.tax %>%
  # re-name Crangon crangon IDs based on rationale above.
  mutate(taxon=ifelse(taxon=="Crangon crangon","Crangon",taxon)) %>%
  mutate(species=ifelse(taxon=="Crangon",NA,species)) %>%
  # FILTER
  filter(taxon %in% indat$species | taxon %in% c("Cottus","Mya","Carcinus","Crangon","Zosteraceae")) %>%
  # fix rank column
  mutate(rank=ifelse(taxon %in% c("Crangon","Cottus","Mya","Carcinus"),"genus","species"))
```

**Save.**
```{r}
if(isFALSE(knitting)){
mc.tax.filter %>% write_csv(here('../','data','results','MockCommunities_BF3_filtered_input_taxa.csv'))
}
```


### adjust taxa for zoid

We want to make sure that when we run zoid, the IDs from the mocks match up to the corresponding IDs in the sample data.



How did the IDs from the mocks / input list show up in the sample dDNA data?
```{r}
unique(mc.tax.filter$taxon)[which(unique(mc.tax.filter$taxon) %in% tax$taxon)]
unique(indat$species)[which(unique(indat$species) %in% tax$taxon)]

```
Unlike the mock community sequencing data, we have *Mya arenaria* IDs in the sample data (as opposed to *Mya*).

There are both *Crangon franciscorum* and *Crangon* IDs in the sample data, but we don't know if *Crangon* is equivalent to *Crangon franciscorum*.

As in the mocks, the sample data has *Zosteraceae* IDs. But we don't want to apply what we know are *Zostera marina* alpha value to the non-specific ID in the sample data.


Which species from the mock community lab list / sequencing data are missing from the sample data?
```{r}
unique(indat$species)[which(!(unique(indat$species) %in% tax$species) & !(unique(indat$species) %in% sample.egc$species))]

unique(mc.tax.filter$taxon)[which(!(unique(mc.tax.filter$taxon) %in% tax$taxon) & !(unique(mc.tax.filter$taxon) %in% sample.egc$taxon))]
```
In the sample data, we don't have *Cottus asper* or *Cottus* IDs.
Again, in the sample data, we don't have *Zostera marina* IDs.


<br>

This means we'll do two runs of the quant model: one with all the species in the mocks to get alpha values, and one with only the species in the sample data to get posterior mean DNA proportions. 

This also means that for the sample run, we want to manually edit *Mya* to *Mya arenaria*, *Crangon* to *Crangon franciscorum*, and *Zosteraceae* to *Zostera marina* in the mocks data that will go into zoid. 



Do this in a new column, 'taxon.species'
```{r}
mc.tax.edit <- mc.tax.filter %>%
  mutate(taxon.species=ifelse(taxon=="Zosteraceae","Zostera marina",
                              ifelse(taxon=="Cottus","Cottus asper",
                                     ifelse(taxon=="Mya","Mya arenaria",taxon))))
```


**In theory,** since we Sanger sequenced the *Crangon franciscorum* DNA that was included in the mocks to verify the species-level visual identification, *Crangon* IDs in the mocks (derived from *Crangon crangon* BLAST output with lower percent identities) should be *Crangon franciscorum*. So we could modify all *Crangon* IDs to *Crangon franciscorum*. But I don't feel super comfortable modifying the data in this way (what if some of the *Crangon* DNA represents tag-jumping from samples from a different *Crangon* species). So I'll include this code here, but not run it.
```{r eval=FALSE}
unique(filter(mc.tax.edit,taxon=="Crangon franciscorum")$sample)  # should be in A,B,D,E
# mc.tax.edit %<>% mutate(taxon.species=ifelse(taxon=="Crangon","Crangon franciscorum",taxon))
```


### summarise reads

mock community: get read counts for all taxa. Use the new column `species.taxon` 
```{r}
mc.reads <- mc.tax.edit  %>%
  filter(taxon.species %in% indat$species) %>%
  mutate(nReads=ifelse(is.na(nReads),0,nReads)) %>%
  group_by(sample,tech,MiSeqRun) %>%
  mutate(techReads=sum(nReads)) %>%
  ungroup() %>%
  mutate(taxon.species = as.character(taxon.species)) %>%
  group_by(sample, taxon.species, tech, genus,techReads,MiSeqRun) %>%
  summarise(taxReads=sum(nReads),
            taxReps=length(unique(tech)),
            taxHashes=length(unique(Hash))) %>%
  mutate(taxReads=ifelse(is.na(taxReads),0,taxReads),
         taxHashes=ifelse(is.na(taxHashes),0,taxHashes)) %>%
   mutate(p.taxReads=taxReads/techReads)
```

Which species in the metadata aren't matched to read counts?
```{r}
unique(filter(mc.reads,is.na(MiSeqRun))$taxon.species)
```
great, none.

Add back in the input information (lab proportions)

```{r}
mc.reads %<>% left_join(indat,by=c("sample","MiSeqRun","taxon.species"="species")) %>%
  filter(!is.na(dna_relative_proportion))
```


#### optional: compare to previous version?

Get the older data set and grab only the columns to match / compare
```{r}
old.mc <- read_csv(here('../../','EGC-Willapa-Bay','data','qm_model','qm_model_MCinput_all-matched_CEGblast.csv'))


old.mc %<>% rename(techReads=totalSpReads) %>% 
  pivot_longer(cols=c(taxReads,taxHashes,p.taxReads,techReads), names_to='compare',values_to='reads_old')

old.mc %<>% dplyr::select(sample,tech,MiSeqRun,taxon.species,sample_label,dna_ng,dna_relative_proportion,compare,reads_old)
```

Join old to new data
```{r}
mc.reads.check <- mc.reads %>%  
  pivot_longer(cols=c(taxReads,taxHashes,p.taxReads,techReads), names_to='compare',values_to='reads_new') %>%
  left_join(old.mc)
```

Look for mismatches
```{r}
mc.mismatch <- filter(mc.reads.check, round(reads_old,4) != round(reads_new,4))
dim(mc.mismatch)
unique(mc.mismatch$compare)
```

#### save

```{r}
if(isFALSE(knitting)){
write_csv(mc.reads,here('../','data','zoid','zoid_MOCK_matchedTaxonomy.csv'))
}
```


### zoid input format

```{r}
mock.a <- mc.reads %>%
  filter(taxon.species %in% indat$species) %>%
  ungroup() %>%
  mutate(CommType=ifelse(grepl("even",sample_label),"Even","Skew")) %>%
  mutate(N_pcr_mock=sum(35,5)) %>%
  rename("species"=taxon.species,"Nreads"=taxReads,"Community"=sample,
         "b_proportion"=dna_relative_proportion,"totReads"=techReads,"propReads"=p.taxReads) %>%
  dplyr::select(Community,CommType,tech,species,Nreads,totReads,propReads,b_proportion,N_pcr_mock)

head(mock.a)
```

insert missing zeros. 
zoid requires zeros any time a species is not present (e.g., green crab is missing from certain mock communities on purpose)
```{r}
dim(mock.a)
mock.a <- mock.a %>% 
  unite(Community, CommType, tech, col = "Sample", remove = T, sep = "#") %>% 
  pivot_wider(names_from = Sample, values_from = Nreads, values_fill = 0) %>% 
  pivot_longer(-c(species, totReads, propReads, b_proportion, N_pcr_mock), names_to = "Sample", values_to = "Nreads") %>% 
  separate(Sample, into = c("Community", "CommType", "tech"), sep = "#")

dim(mock.a)
```



## Diet data prep

recall: we have the sample blast taxonomy that was filtered / combined, and the sample green crab reads
```{r}
colnames(sample.tax)
colnames(sample.egc)
```


### final clean-up

grab only the columns of interest from the blast data. make sure that the "sample_id" column is present and formatted correctly.
```{r}
sample.tax %<>% mutate(tech_abc=letters[tech]) %>% unite(col="sample_id",sample,tech_abc,sep="-", remove=FALSE) %>% dplyr::select(-tech_abc)

sample.tax.thin <- sample.tax %>%
  dplyr::select(sample_id,sample,tech,Hash,nReads,taxon,species,genus,family,rank,MiSeqRun)
```

correct the 'sample_id' in the egc data table to matched the corrected technical replicate numbers (for samples run across two miseq runs).
```{r}
sample.egc %<>%
  dplyr::select(-sample_id) %>% 
  mutate(tech_abc=letters[tech]) %>%
  unite(col="sample_id",sample,tech_abc,sep="-",remove=FALSE)
```

check to make sure the EGC data doesn't have any more duplicated samples between runs 3 and 4
```{r}
sample.egc.run3 <- filter(sample.egc,MiSeqRun==3) %>% dplyr::select(sample,tech,sample_id,MiSeqRun) %>% distinct()
sample.egc.run4 <- filter(sample.egc,MiSeqRun==4) %>% dplyr::select(sample,tech,sample_id,MiSeqRun) %>% distinct()
sample.egc.shared <- bind_rows(sample.egc.run3,sample.egc.run4) %>%
  dplyr::select(-sample_id) %>% distinct() %>%
  group_by(sample,tech) %>%
  summarise(nruns=length(unique(MiSeqRun))) 
sample.egc.shared %>% filter(nruns > 1)
```
All good!

### Add in the green crab hashes 

```{r}
sample.tax.thin %<>% bind_rows(sample.egc %>% dplyr::select(colnames(sample.tax.thin)))

any(is.na(sample.tax.thin$sample))
```

Check to make sure the combined EGC / sample data doesn't have duplicated samples between runs
```{r}
to_edit <- sample.tax.thin %>% group_by(sample, tech) %>% 
  summarise(nruns=length(unique(MiSeqRun)), runs=paste0(unique(MiSeqRun),collapse=",")) %>% filter(nruns > 1)
to_edit
```


deal with technical replicates where adding the *Carcinus maenas* taxonomy back into the dataset created duplicate sample IDs between runs -- by adjusting technical replicates in run 3 manually
```{r}
to_edit <- to_edit %>%
  left_join(sample.tax.thin,by=c("sample","tech"))
# View(to_edit)

keep106 <- filter(sample.tax.thin, sample=="WACO21-106") %>%
  filter(!(MiSeqRun==3 & taxon %in% c("Carcinus maenas","Carcinus")))
  
tmp106.run3 <- filter(sample.tax.thin, sample=="WACO21-106") %>%
  filter(MiSeqRun==3 & taxon %in% c("Carcinus maenas","Carcinus")) %>%
  mutate(tech=tech+6) %>%
  mutate(tech_abc=letters[tech]) %>%
  dplyr::select(-sample_id) %>%
  unite(col="sample_id",sample,tech_abc,sep="-", remove=FALSE) %>%
  dplyr::select(-tech_abc)

# replace the old 106 technical replicate numbers
sample.tax.thin %<>% filter(sample != "WACO21-106") %>% 
  bind_rows(tmp106.run3) %>% bind_rows(keep106)

# check again
sample.tax.thin %>% group_by(sample, tech) %>% 
  summarise(nruns=length(unique(MiSeqRun)), runs=paste0(unique(MiSeqRun),collapse=",")) %>% filter(nruns > 1)
```


There are no other *Carcinus* species in the region, so we can assume that any genus-level IDs for *Batillaria* match the *Carcinus maenas* we put in the mocks. Specify any genus-level IDs in the sample data.
```{r}
filter(sample.tax.thin, taxon=="Carcinus")
```
Great, this doesn't apply.


Let's also check for non-specific reads from other groups that don't have other species in their genera in the region:
```{r}
check_genus <- c("Mya","Leptocottus","Metacarcinus","Carcinus","Ruditapes","Batillaria","Cymatogaster")
filter(sample.tax.thin,genus %in% check_genus & is.na(species))
```
All good!

Oops, remove controls
```{r}
sample.tax.thin %<>% filter(!(grepl("Control",sample)))
```

#### save

```{r}
if(isFALSE(knitting)){
write_csv(sample.tax.thin,here('../','data','zoid','zoid_SAMPLEdata_taxonomy.csv'))
}
```


### N PCR cycles
Read in the metadata file that contains the number of PCR cycles each sample was run for, in the final metabarcoding dataset
```{r fig.height=4, fig.width=4}
n_pcr_samples <- read_csv(here('../','data','metadata','Willapa Bay EGC Samples - PCR cycles.csv'))

Obs_PCRcycles <- sample.tax.thin %>%
  dplyr::select(sample_id,sample,tech,MiSeqRun) %>%
  distinct() %>%
  left_join(n_pcr_samples,by=c("sample_id"="sample","MiSeqRun"))

ggplot(Obs_PCRcycles, aes(x=PCRcycles,fill=as.factor(MiSeqRun)))+ geom_bar() + theme_bw()
```

Make sure no samples have duplicates / more than one pcr cycle value (want this to be true)
```{r}
dim(sample.tax.thin %>%
  dplyr::select(sample_id,sample,tech,MiSeqRun) %>%
  distinct())[1] == dim(Obs_PCRcycles)[1]
```
(Want this to be 0)
```{r}
Obs_PCRcycles %>% group_by(sample_id) %>% summarise(n=n()) %>% filter(n>1)
```


Save
```{r}
if(isFALSE(knitting)){
write_csv(Obs_PCRcycles,here('../','data','zoid','zoid_SAMPLEdata_PCR.csv'))
}
```


### qm input format

```{r}
metabarcoding <- sample.tax.thin %>%
  rename("Sample"=sample,"Nreads"=nReads) %>%
  mutate(station=Sample,
         time=1,creek=1) %>%
  group_by(Sample,station,creek,time,tech,taxon) %>%
  summarise(Nreads=sum(Nreads)) %>%
  rename("species"=taxon)
```


insert missing zeros
```{r}
dim(metabarcoding)
metabarcoding %<>% 
  unite(station, creek, time, tech, col = "Sample", remove = T, sep = "#") %>% 
  pivot_wider(names_from = Sample, values_from = Nreads, values_fill = 0) %>% 
  pivot_longer(-c(species), names_to = "Sample", values_to = "Nreads") %>% 
  separate(Sample, into = c("station", "creek", "time", "tech"), sep = "#")
dim(metabarcoding)
```




### optional: compare to old data
```{r eval=FALSE}
load(file=here('../','data','zoid','qm_model_dataLIST_2023-8-29.rds')); old.dat <- EGC_obs[[2]]
rm(EGC_obs)
```
```{r eval=FALSE}
check.dat <- metabarcoding %>% rename(Nreads_new=Nreads) %>%
 full_join(old.dat %>% rename(Nreads_old=Nreads) %>% mutate(species=ifelse(species=="Mya","Mya arenaria",species))) 
View(check.dat %>%
  filter(Nreads_new != Nreads_old | is.na(Nreads_old) | is.na(Nreads_new)))
```



# Save QM input

## sample run

```{r}
EGC_obs <- list(mock.a,metabarcoding)
names(EGC_obs) <- c("mock","metabarcoding")

head(mock.a)
head(metabarcoding)

unique(intersect(mock.a$species, metabarcoding$species))
mock.a$species[which(!(mock.a$species %in% metabarcoding$species))]
```

Save: the metabarcoding_data object, and the mock / metabarcoding dataframes in a list
```{r}
if(isFALSE(knitting)){
save(EGC_obs,file=here('qm_model_dataLIST_2023-8-29.rds'))
}
```



## alpha run

For this run, keep the 
```{r}
```

```{r}
EGC_obs2 <- list(mock.a,metabarcoding)
names(EGC_obs2) <- c("mock","metabarcoding")

head(mock.a)
head(metabarcoding)

unique(intersect(mock.a$species, metabarcoding$species))
mock.a$species[which(!(mock.a$species %in% metabarcoding$species))]
```

Save: the metabarcoding_data object, and the mock / metabarcoding dataframes in a list
```{r}
if(isFALSE(knitting)){
save(EGC_obs2,file=here('qm_model_dataLIST_matchAlpha_2023-8-29.rds'))
}
```


